[BUG] - Handle deserialization of arrays of nested objects

# Handle deserialization of arrays of nested objects

Allow something like fromJsonAs() for arrays of nested objects. Related to #62, but for deserialization. For example:

```ts
class Nested extends Serializable {
    @SerializeProperty("a_property")
    test = true;
}

class Outer extends Serializable {
    @SerializeProperty({ fromJsonStrategy: fromJsonAs(Nested) })
    outer: Nested[] = [Nested(), Nested()];
}
```

Currently this fails:
```ts
test({
  name: "fromJsonAs works with arrays of objects",
  fn() {
    class Test extends Serializable {
      @SerializeProperty("a_property")
      test = true;
    }
    const testObj = new Test();
    const array: Test[] = fromJsonAs(Test)([{ a_property: "v1" }, { a_property: "v2" }]);
    assertEquals(array.length, 2);
    assert(array[0] instanceof Test);
  },
});
```
```
failures:

fromJsonAs works with arrays of objects
AssertionError: Values are not equal:


    [Diff] Actual / Expected


-   undefined
+   2
```

## Motivation
Make it easier to handle arrays of nested objects.

## Design Detail
Make fromJsonAs() handle arrays by checking if the input is an array before it tries to build a new object.

## Drawbacks
- ...

## Rationale and Alternatives
Alternative: add a separate fromJsonAsArray() that callers can use instead of fromJsonAs(), something like this:

```ts
export function fromJsonAsArray<T>(type: T & { new (): Serializable }): FromJsonStrategy {
      return (value: any) => (value ? value.map((element: any) => new type().fromJson(element)) : null);
}
```

But I'm not sure the separate function adds any value. It's just one more thing callers would have to remember to do.

## Prior Art


## Unresolved questions

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - deno@1.5.1
 - std@0.76.0
 - fmt changes with deno upgrade
+- fixed #64
 
 ### Added
 - interface `tsTransformKey`

--- a/from_json/as.ts
+++ b/from_json/as.ts
@@ -6,5 +6,10 @@ import { FromJsonStrategy, Serializable } from "../serializable.ts";
 export function fromJsonAs<T>(
   type: T & { new (): Serializable },
 ): FromJsonStrategy {
-  return (value: T) => new type().fromJson(value);
+  return (value: T) => {
+    if (Array.isArray(value)) {
+      return value.map((item) => new type().fromJson(item));
+    }
+    return new type().fromJson(value);
+  };
 }

--- a/from_json/as_test.ts
+++ b/from_json/as_test.ts
@@ -47,3 +47,19 @@ test({
     assertEquals(testObj.test2.test1, false);
   },
 });
+
+test({
+  name: "fromJsonAs works with arrays of objects",
+  fn() {
+    class Test extends Serializable {
+      @SerializeProperty("a_property")
+      test = true;
+    }
+    const array: Test[] = fromJsonAs(Test)(
+      [{ a_property: "v1" }, { a_property: "v2" }],
+    );
+    assertEquals(array.length, 2);
+    assert(array[0] instanceof Test);
+    assertEquals(array[0].test, "v1");
+  },
+});

